Bug Fixes for the new RTC/Sensors driver maxim,ds3231

rexut opened the pull request:
- avoid warnings about unused code (in case of unused features, e.g. alarm or update)
- avoid compilation error (probably most important for the codebase, a really unsightly typo)
- as the C language treats “cases” similarly to “labels”, a “case” block must begin with a statement and not with a variable declaration (declarations are not statements)

rexut committed:
drivers: rtc: maxim,ds3231: avoid warnings about unused code

Compile warnings occurred in common standard use case
w/o RTC alarm and update.

Warning was:

drivers/rtc/rtc_ds3231.c: In function 'rtc_ds3231_init':
drivers/rtc/rtc_ds3231.c:808:33: warning: unused variable 'data'
                 [-Wunused-variable]
  808 |         struct rtc_ds3231_data *data = dev->data;
      |                                 ^~~~
drivers/rtc/rtc_ds3231.c: At top level:
drivers/rtc/rtc_ds3231.c:195:12:
        warning: 'rtc_ds3231_get_ctrl_sts' defined but not used
                 [-Wunused-function]
  195 | static int rtc_ds3231_get_ctrl_sts(const struct device *dev,
      |                                    uint8_t *buf)
      |            ^~~~~~~~~~~~~~~~~~~~~~~

Signed-off-by: Stephan Linz <linz@li-pro.net>

rexut committed:
drivers: rtc: maxim,ds3231: avoid compilation error

Because of wrong integer type syntax following error occurred:

drivers/rtc/rtc_ds3231.c:361:76: error: unknown type name 'u'
  361 | static int rtc_ds3231_alarm_get_supported_fields(
      |                    const struct device *dev, u int16_t id,
      |                                              ^

Signed-off-by: Stephan Linz <linz@li-pro.net>

gergovari commented:
> @gergovari, @bjarki-andreasen and @ttmut FYIO, some bug fixes – maybe still to be merged for Zephyr v4.1?

Thank you for fixing it up!

rexut commented:
@bjarki-andreasen and @cfriedt: many thx for fast review, but please wait one day more with merge. Just I stumbled over another compile error in the `maxim,ds3231-sensor` driver for the digital (ambient) temperature:

```
drivers/sensor/maxim/ds3231/ds3231.c: In function ‘sensor_ds3231_channel_get’:
drivers/sensor/maxim/ds3231/ds3231.c:77:3: error: a label can only be part of a statement and a declaration is not a statement
   77 |   const uint16_t raw_temp = data->raw_temp;
      |   ^~~~~
```

I'll have a look on it too, maybe with a fix as third commit in this PR in the next few days.

rexut committed:
drivers: sensor: maxim,ds3231: avoid compilation error

As the C language treats 'cases' similar to 'labels' the following
error can also occur when using switch statements when it encounters
a declaration immediately after a label:

drivers/sensor/maxim/ds3231/ds3231.c:77:3: error: a label can only
        be part of a statement and a declaration is not a statement
   77 |   const uint16_t raw_temp = data->raw_temp;
      |   ^~~~~

Signed-off-by: Stephan Linz <linz@li-pro.net>

bjarki-andreasen commented:
You can move the PR back to draft to prevent it being merged :)

rexut commented:
> You can move the PR back to draft to prevent it being merged :)

Yep, I know. On the end, it was a quick fixup. Now it's ready to review.

bjarki-andreasen commented on drivers/sensor/maxim/ds3231/ds3231.c:
```diff
@@ -73,7 +73,7 @@ static int sensor_ds3231_channel_get(const struct device *dev, enum sensor_chann
 	switch (chan) {
-	case SENSOR_CHAN_AMBIENT_TEMP:
+	case SENSOR_CHAN_AMBIENT_TEMP: {
 		const uint16_t raw_temp = data->raw_temp;
```
Variables need to be declared in the beginning of the function, right under `struct sensor_ds3231_data *data = dev->data;`, otherwise, move out the whole calculation into a helper:
```
static void raw_temp_to_sensor_value()
{
        ...
}

I prefer helpers in any case :)

...

switch (chan) {
case SENSOR_CHAN_AMBIENT_TEMP:
        raw_temp_to_sensor_value();
        break

...
```

rexut replied:
Mine as well, I'm just surprised that this issue wasn't noticed and reported in the first review. However, I will change this for the 'get' function.

Nevertheless, @gergovari should go through all MFD/RTC/SENS artifacts of the driver again and submit a clean-up PR accordingly, if this is such a hard requirement of the Zephyr Code Guidelines.

cfriedt commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -805,7 +808,9 @@ static int rtc_ds3231_init(const struct device *dev)
 	int err = 0;
 
 	const struct rtc_ds3231_conf *config = dev->config;
+#if defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM)
 	struct rtc_ds3231_data *data = dev->data;
+#endif
```
The ifdefery can be reduced with `__maybe_unused` (although the position of it might need to be adjusted).

```suggestion
  __maybe_unused
 	struct rtc_ds3231_data * data = dev->data;
```

rexut replied:
Hey, good idea, at least for local variables. I'll do so and roll out a reworked patch set. For static functions I'm just unsure, if the compiler or linker optimization would skip out the symbols to reduce memory consumption and avoid dead code. Thus I would still ifdef-out them.

cfriedt commented on drivers/sensor/maxim/ds3231/ds3231.c:
```diff
@@ -37,6 +37,16 @@ struct sensor_ds3231_conf {
+	uint8_t frac = raw_temp & 3;
+
+	val->val1 = (int8_t)(raw_temp & GENMASK(8, 2)) >> 2;
+	val->val2 = (frac * 25) * pow(10, 4);
```
Would be ideal to avoid the use of floats altogether here. Can we use an integer literal here rather than pulling in double precision floating point?

rexut replied:
I would like to give this decision back to @gergovari. Such a change goes far beyond the intention of this bug fixing PR, doesn't it? I think a new PR with more functional improvements is the right approach here.

MaureenHelm replied:
Agree it would be nice to avoid floats, but since it was already here it's ok to leave it.

rexut replied:
I will not change any code algos, interfaces or behaviors in a bug fix PR. @gergovari could you please pick up any comments for improvements and QA validation (regression testing) and submit the necessary changes with an improvement PR? Thank you very much.

MaureenHelm commented on drivers/sensor/maxim/ds3231/ds3231.c:
```diff
@@ -74,13 +84,7 @@ static int sensor_ds3231_channel_get(const struct device *dev, enum sensor_chann
 
 	switch (chan) {
 	case SENSOR_CHAN_AMBIENT_TEMP:
-		const uint16_t raw_temp = data->raw_temp;
-
-		val->val1 = (int8_t)(raw_temp & GENMASK(8, 2)) >> 2;
-
-		uint8_t frac = raw_temp & 3;
-
-		val->val2 = (frac * 25) * pow(10, 4);
+		sensor_ds3231_temp_from_raw(data, val);
```
You can wrap the block in curly braces instead of pulling out to a separate function.

rexut replied:
This has already been clarified, see: https://github.com/zephyrproject-rtos/zephyr/pull/85030#discussion_r1938898404

For MISRA rule reasons, statement blocks are considered “undesirable”, so there is now a static inline function for this (also my preferred variant).

MaureenHelm reviewed (commented):
Thanks for the fix. Would you mind adding the sensor to `tests/drivers/build_all/sensor/spi.dtsi`? This will ensure it builds in CI and help keep it from breaking again.

rexut commented:
> Thanks for the fix. Would you mind adding the sensor to `tests/drivers/build_all/sensor/spi.dtsi`? This will ensure it builds in CI and help keep it from breaking again.

I would like to give this requirement back to @gergovari. Such a change goes far beyond the intention of this bug fixing PR, doesn't it?

I am well aware of the need for tests (at least regression tests). However, all these wishes (or requirements) were not requested in the initial review and are therefore not available. I see all of these things as a list of correct comments, but they should be collected separately by the maintainers via an FRC issue and submitted to a review for processing via separate improvement PRs.

rexut commented:
> FRC issue?

Sorry, my typo. Should be "RFC issue" (Request for Comments) as an starting point for all the current requests (changes we now to do) and further (currently unknown points) where people can have a kind of planning platform so that a review with improvements (PR) can finished smoothly. But now, the time I'm writing this, I see an RFC is overloaded here. That is more for 'treewide' change requests.

As the simplest variant, you can also simply use the “Enhancement” template when creating a new issue (on GitHub) and collect all the points first. Discussion is also possible.
